[12.0][IMP] fieldservice_sale

brian10048 opened the pull request:
#393

Depends on:
- [x] #411

brian10048 committed:
[IMP] fieldservice_sale: Add service policy

- This change will allow to create an FSM order for each sales order and link it to the delivery order and invoice

brian10048 committed:
[ADD] fieldservice_sale_stock

hparfr commented on fieldservice_sale/models/sale_order.py:
```diff
@@ -25,10 +25,76 @@ class SaleOrder(models.Model):
         for order in self:
-            order.fsm_order_ids = self.env['fsm.order'].search([
+            orders = self.env['fsm.order']
+            orders += self.env['fsm.order'].search([
```
I think it's better to use |= because it's an union of recordsets.

brian10048 replied:
Thank you.  Updated in a recent commit

hparfr commented on fieldservice_sale/models/sale_order.py:
```diff
@@ -25,10 +25,76 @@ class SaleOrder(models.Model):
+
+    @api.multi
+    def _field_find_fsm_order(self):
+        """ Find the fsm_order generated by the Sale Order. If no fsm_order
```
I think it's better to adapt the same behavior as delivery orders : when you cancel the sale, the picking is cancelled, when set to quotation then confirmed again a new picking is created.

brian10048 replied:
That is probably a better method than this is being done here.  I agree there is a fair amount of code duplication between sale_order and sale_order_line.  While similar they are also a little different.

I originally modeled this after v11 of sale_timesheet.

hparfr reviewed (changes requested):
There is a lot of code duplication between sale_order and sale_order_line about creating fsm_orders.

brian10048 committed:
[IMP] fieldservice_sale: Link invoices to FS order

max3903 committed:
[IMP] fieldservice_sale: Merge and reuse fields

max3903 committed:
[IMP] fieldservice_sale_recurring

max3903 commented:
@brian10048 Nice job. Please see https://github.com/brian10048/field-service/pull/6

brian10048 committed:
[IMP] fieldservice_sale

- Combine field service template info from sale order lines and add it to field service order

brian10048 commented:
@max3903 I made the additions to copy template information to the field service order

hparfr commented on fieldservice_sale/__manifest__.py:
```diff
@@ -11,6 +11,7 @@
     'depends': [
         'fieldservice',
         'sale_management',
+        'fieldservice_account',
```
Please don't add this dependency.

brian10048 replied:
@hparfr I agree adding the dependency adds a lot of extra logic from fieldservice_account that is not needed for this addition, but if this dependency is not used, there is no `fsm_order_id` field for `account.invoice` model which we set here from the `sale_order`

Thoughts?

max3903 replied:
@hparfr `sale_management` depends on `account`.

If a FSM product is set to be invoiced on quantity delivered, we need the ability to invoice the completed FSM order.

hparfr replied:
Yes, but fieldservice_account take strong assumption about how invoicing should be done
and some dependencies (analytic, hr_timesheet)

max3903 replied:
How would you do it?

brian10048 replied:
Can we split fieldservice_account so that the way it does invoicing is in a separate module and keep only things like adding fsm_order_id to the account.invoice model?
I don't mind doing the work

brian10048 replied:
@max3903 @hparfr Take a look at #411 and please provide your feedback

hparfr replied:
Thanks for this work @brian10048.
I still think billing_id (customer_id) and costs should not be in fsm_order. There is not so much stuff left in fieldservice_accounting but just the link between account_invoice_line and fsm_order

IMO fieldservice_sale should: 
- link fsm_orders to sale_orders (ok)
- choose how fsm_orders are created (ok field_service_tracking)
- create fsm_orders from sale (ok)
- compute invoiceable quantity (ok)
- prepare invoice lines (ok)

max3903 replied:
@hparfr I am fine with this. On the last 2 items (compute invoiceable quantity and prepare invoice lines), if you invoice on delivered quantity, you would invoice from the sale order, right? Not from the FSM order (like inventory items are not invoiced from the delivery order).

# fieldservice_sale

* Invoice the service from the sale order before or after the completion of the service (--> no timesheet, no cost)

# fieldservice_account

* Invoice timesheet/costs from the FSM order

# fieldservice_sale_account

* Invoice the service + timesheet/costs from either the FSM order or the sale order, before (no timesheet/cost) or after the completion of the service

hparfr replied:
Indeed @max3903 it's like sale orders and pickings:  _compute_qty_delivered() is currently in fieldservice_sale/models/sale_order_line.py

In the actual proposal
- fieldservice_account: links a fsm_orders to invoices lines. Kind of base module.
- fieldservice_account_analytic:  add timesheet, costs, billing_id, and so on. Depends of fieldservice_account
- fieldservice_sale, create fsm orders at sale confirmation and set quantities when invoice is created from sale. Depends of fieldservice_account
- fieldservice_contract, set quantities when invoice is created from contract. Depends of fieldservice_account

max3903 replied:
:+1:

max3903 replied:
@hparfr @brian10048 See #415 where I copied `fieldservice_account` and removed the `fsm_order_id` field on `account.invoice.line`

hparfr replied:
@max3903  have you seen https://github.com/brian10048/field-service/pull/8 ? 
I will look at #415 in few days.

brian10048 replied:
I will be able to look at all of this over the next few days as well and provide feedback on all of the above

brian10048 replied:
Resolved with #411

hparfr reviewed (changes requested):
Accounting should not be a dependency of this module.

max3903 commented:
@brian10048 The Field Service tracking logic seems to only apply to service product. I need it to apply to stockable product as well.

I created 4 stockable products: 2 tracked by sale, 2 tracked by line and got no FSM order when confirming the sale order.

My scenario is that I am delivering stockable products (carpet or water bottle) and both are considered delivered when:
* the stock picking is transferred
* the invoice is paid
* the FSM order is completed (product is installed, customer signed)

Edit: Thinking about this, maybe we should add a column on the sale order line "Qty Serviced" and not touch the "Qty Delivered" logic.
The quantity delivered would still be computed based on the stock moves.
The quantity serviced would be computed based on the completion of the related FSM order.
It would be clear on the sale order to see whether or not an item has been delivered or not, and installed or not.
Some organizations could invoice based on the quantity delivered or based on the quantity delivered and serviced.

What do you think?

CC: @wolfhall

brian10048 commented:
@max3903 Did you have fieldservice_sale_stock module installed?  I don't think it is in the settings page.  It should provide the functionality you mentioned needing for creating the field service order with a stockable product.  I separated it to help minimize dependencies of fieldservice_sale

I'm not opposed to a qty_serviced field for the sale order line

bodedra commented on fieldservice_sale_stock/__manifest__.py:
```diff
@@ -0,0 +1,23 @@
+
+{
+    'name': 'Field Service - Sale Stock',
+    'version': '12.0.0.1.0',
```
```suggestion
    'version': '12.0.1.0.0',
```

bodedra reviewed (approved):
Code review. :+1:

hparfr commented on fieldservice/models/res_config_settings.py:
```diff
@@ -84,6 +86,14 @@ def _onchange_module_fieldservice_stock(self):
+    def _onchange_module_fieldservice_sale(self):
+        if self.module_fieldservice_sale and \
+           self.module_fieldservice_stock:
+            self.module_fieldservice_sale_stock = True
```
Use auto_install = True and depends in fieldservice_sale_stock/\_\_manifest\_\_.py instead of the _onchange.

brian10048 replied:
Thank you! Updated in latest commit

brian10048 committed:
[IMP] fieldservice_sale_stock: Auto Install module

hparfr commented:
Thanks @brian10048 
IMO we can merge now

brian10048 commented:
From @max3903 previous comment:
> @brian10048 The Field Service tracking logic seems to only apply to service product. I need it to apply to stockable product as well.
> 
> I created 4 stockable products: 2 tracked by sale, 2 tracked by line and got no FSM order when confirming the sale order.
> 

I thought this was related to not having fieldservice_sale_stock module installed, but I am running into the same issue on runbot.  I remember it was working as intended in previous testing, so I am looking to see what got changed and get this working correctly before merge

brian10048 committed:
[FIX] fieldservice_sale

Consumable and Storable product types were not generating field services

brian10048 commented:
@max3903 @hparfr @bodedra Any suggestions that might point me in the right direction to fix this?

I rebased my branch to reflect any recent changes to fieldservice_account module, but this did not fix the problem.

When testing on a local Enterprise install, the pickings from stock products are properly linked to a newly created FSM order, but I am not receiving same result on the runbot Community version.

Both versions work as intended for a product type set to service. 

# Enterprise
## FSM Tracking Set to per Sale Order
![image](https://user-images.githubusercontent.com/25710115/70341401-d7fe2a80-1820-11ea-8484-a8f812267e9a.png)

## Confirmed SO
Notice the linked FSM order smart button
![image](https://user-images.githubusercontent.com/25710115/70341127-40004100-1820-11ea-926e-9ee2987a50c6.png)

## Pickings linked to FSM Order
![image](https://user-images.githubusercontent.com/25710115/70341782-b6517300-1821-11ea-9b0b-63cf12d0ad0d.png)



# Community
## FSM Tracking Set to per Sale Order
![image](https://user-images.githubusercontent.com/25710115/70341463-f6fcbc80-1820-11ea-9cd1-20a729ed30f7.png)

## Confirmed SO
No FSM Order linked
![image](https://user-images.githubusercontent.com/25710115/70341144-4c849980-1820-11ea-97c9-967f2808c83e.png)

max3903 commented on fieldservice_sale/readme/CONFIGURE.rst:
```diff
@@ -1,12 +1,27 @@
 
 * Go to Sales > Catalog > Products
 * Create or select a product
 * Set the Product Type to 'Service' under General Information tab
```
```suggestion
```

brian10048 committed:
[UPD] fieldservice_sale: readme

max3903 commented:
@brian10048 As we use templates on the product, the "Manager Order Templates" option should be activated in Field Service.

max3903 commented:
@brian10048 I still can't get any FSM orders to be created:
http://3400685-403-2ed3cb.runbot1.odoo-community.org/web?#id=23&action=480&model=sale.order&view_type=form&menu_id=329

Product 1 & 2 are set to create 1 FSM order per sale order.
Product 3 & 4 are set to create 1 FSM order per sale order line.

What am I missing?

brian10048 commented:
@max3903. I'm having the same problem on runbot, but not when I'm testing on an Enterprise version.

Can you help me figure out what I'm missing?

max3903 commented:
@brian10048 Sure. I will debug this tomorrow morning.

max3903 committed:
[FIX] fieldservice_sale: Missing the template info

max3903 commented:
@brian10048 Please review https://github.com/brian10048/field-service/pull/9

I believe the issue is with the `fieldservice_sale_recurring` module. I was not able to reproduce the issue on CE and EE when it is not installed.

brian10048 committed:
Merge pull request #9 from ursais/dev-sale-policy

[FIX] fieldservice_sale: Missing the template info

brian10048 commented:
> 
> @brian10048 Please review [brian10048#9](https://github.com/brian10048/field-service/pull/9)
> 
> I believe the issue is with the `fieldservice_sale_reccuring` module. I was not able to reproduce the issue on CE and EE when it is not installed.

OK, thank you! That has been driving me crazy. 
Makes sense because I never installed fieldservice_sale_recurring on my EE test.

Should I look into and fix the fieldservice_sale_recurring module as part of this PR or should we get this merged and work on that separately?

max3903 committed:
[FIX] fieldservice_sale: Missing location directions

max3903 committed:
[FIX] fieldservice_sale_stock: Expecting one result

max3903 commented:
@brian10048 Let's merge this one first.

I made another pull request: https://github.com/brian10048/field-service/pull/10

brian10048 committed:
Merge pull request #10 from ursais/dev-sale-policy

[FIX] fieldservice_sale: Missing location directions

brian10048 commented:
Thanks @max3903 

I can work on fixing `fieldservice_sale_recurring` over the next few days as time permits.

max3903 committed:
[ADD] fieldservice_sale: Migration script to 12.0.2.0.0

max3903 commented:
@brian10048 Here is the migration script: https://github.com/brian10048/field-service/pull/11

brian10048 committed:
Merge pull request #11 from ursais/dev-sale-policy

[ADD] fieldservice_sale: Migration script to 12.0.2.0.0

max3903 commented:
/ocabot merge
